src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/TensorPrimitives.netcore.cs

```diff
@@ -0,0 +1,161 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using System.Runtime.Intrinsics;
+
+namespace System.Numerics.Tensors
+{
+    /// <summary>Performs primitive tensor operations over spans of memory.</summary>
+    public static partial class TensorPrimitives
+    {
+        /// <summary>Computes the element-wise result of: <c><paramref name="x" /> + <paramref name="y" /></c>.</summary>
+        /// <param name="x">The first tensor, represented as a span.</param>
+        /// <param name="y">The second tensor, represented as a span.</param>
+        /// <param name="destination">The destination tensor, represented as a span.</param>
+        /// <exception cref="ArgumentException">Length of '<paramref name="x" />' must be same as length of '<paramref name="y" />'.</exception>
+        /// <exception cref="ArgumentException">Destination is too short.</exception>
+        /// <remarks>This method effectively does <c><paramref name="destination" />[i] = <paramref name="x" />[i] + <paramref name="y" />[i]</c>.</remarks>
+        public static unsafe void Add(ReadOnlySpan<float> x, ReadOnlySpan<float> y, Span<float> destination)
+        {
+            if (x.Length != y.Length)
+            {
+                ThrowHelper.ThrowArgument_SpansMustHaveSameLength(nameof(x), nameof(y));
+            }
+
+            if (x.Length > destination.Length)
+            {
+                ThrowHelper.ThrowArgument_DestinationTooShort();
+            }
+
+            ref float xRef = ref MemoryMarshal.GetReference(x);
+            ref float yRef = ref MemoryMarshal.GetReference(y);
+            ref float dRef = ref MemoryMarshal.GetReference(destination);
+            int i = 0, oneVectorFromEnd;
+
+#if NET8_0_OR_GREATER
+            if (Vector512.IsHardwareAccelerated)
+            {
+                oneVectorFromEnd = x.Length - Vector512<float>.Count;
+                while (i <= oneVectorFromEnd)
+                {
+                    Vector512<float> sum = Vector512.LoadUnsafe(ref xRef, (uint)i) + Vector512.LoadUnsafe(ref yRef, (uint)i);
+                    Vector512.StoreUnsafe(sum, ref dRef, (uint)i);
+
+                    i += Vector512<float>.Count;
+                }
+            }
+#endif
+
+            if (Vector256.IsHardwareAccelerated)
```

Comment:
> Ideally we'd get a trailing mask and merge, then store the final result instead

If you can share an example of what that looks like, I can change it.

Reply:
Noting that if we're not supporting overlapping src/dst buffers, we can just backtrack by `remainder` and do 1x full Vector128/256/512 operation, since `Add` is idempotent

otherwise, we want something like: https://github.com/dotnet/machinelearning/blob/d6927515d8014442d35ee22817ec7c40b9e7eec7/src/Microsoft.ML.CpuMath/AvxIntrinsics.cs#L459-L564

which is an example in ML.NET that handles alignment and remaining data. It was written back around .NET Core 3.1, but the general premise should apply equally to the xplat APIs.

Basically we need a small upfront table covering the 4/8/16 different possible leading/trailing masks: https://github.com/dotnet/machinelearning/blob/d6927515d8014442d35ee22817ec7c40b9e7eec7/src/Microsoft.ML.CpuMath/AvxIntrinsics.cs#L24-L46

and we backtrack by `remainder` so that we do 1 last operation that exactly fills a Vector128/256/512, simply doing a conditional select of the already computed results with the new results.

----------------------------

Particularly when processing large data (as may be expected for tensors), handling alignment is important to avoid known perf penalties. We don't strictly have to pin the data, we can "opportunistically align" under the presumption the GC won't move the data by checking the base alignment of the underlying pointer. This will perform well if the GC doesn't move the data and remain correct even if it does (just no longer with the benefit of alignment).

For small data, handling "remaining" elements (that doesn't fit into a vector) efficiently is important and its pretty easy to do. Ideally we add some official APIs to help with this, which is part of the broader consideration around efficient masking support that has already been getting discussed alongside AVX-512, SVE, and general existing ISAs.

Reply:
> Noting that if we're not supporting overlapping src/dst buffers

Are we?

Reply:
I think most of the other APIs we've added to the BCL don't support them.

We should probably double check with @luisquintanilla / @michaelgsharp on whether ML.NET or SK need it.

Reply:
I think for the case of ML.NET that @tannergooding is right about just having a 1x `Span<T>` API when we want it that way. At least thats what it looks like to me.